src/main/scala/com/snowplowanalytics/kinesistee/operators.scala

```diff
@@ -0,0 +1,91 @@
+
+import scalaz.syntax.validation._
+import scalaz.{Failure, Success}
+import scalaz.ValidationNel
+
+/**
+  * Operators work with ValidationNel[Throwable, Content]
+  * and return ValidationNel[Throwable, Content] allowing
+  * chaining.
+  */
+trait Operator extends Product with Serializable {
+  def apply(content: ValidationNel[Throwable, Content]): ValidationNel[Throwable, Content]
+}
+
+/**
+  * Javascript Operator that can be used as a filter or transformer
+  * @param js
+  */
+case class JavascriptOperator(js: String) extends Operator {
+
+  val engine = new ScriptEngineManager(null).getEngineByName("nashorn")
+  if (engine==null) { throw new IllegalStateException("Nashorn script engine not available")}
+  val in: Invocable = engine.asInstanceOf[Invocable]
+  engine.eval(new StringReader(js))
+
+  /**
+    * JavaScript filter - invokes nashorn on the given js script
+    * The script must contain a 'operate' function that accepts the record as an argument
+    * @param content the record to operate on
+    * @return Either NonEmptyContent or FilteredContent
+    */
+
+  override def apply(content: ValidationNel[Throwable, Content]): ValidationNel[Throwable, Content] = {
+    content match {
+      case Success(NonEmptyContent(row, partitionKey)) =>
+        try {
+          val retVal = in.invokeFunction("operator", row)
+          retVal match {
+            case row: String => NonEmptyContent(row, partitionKey).success
+            case bool:java.lang.Boolean => if (bool) NonEmptyContent(row, partitionKey).success else FilteredContent.success
+            case e => new RuntimeException(s"'$e' returned by your js function cannot be converted to a row").failureNel
+          }
+        } catch {
+          case e @ (_: ScriptException | _: NoSuchMethodException ) => e.failureNel
+        }
+      case Success(FilteredContent) => FilteredContent.success
+      case Failure(f) => throw new IllegalStateException(s"Preceding operation has failed '$content': ${f.head}")
+    }
+  }
+}
+
+
+case class SnowplowEnrichedToNestedJsonTransformOperator() extends Operator {
+  /**
+    * Use the Snowplow Scala analytics SDK to turn an enriched event into nested JSON
+    * @param content the record to transform
+    * @return a nested json representation of the enriched event, or failure if it could not be converted
+    */
+  def apply(content: ValidationNel[Throwable, Content]): ValidationNel[Throwable, Content] = {
+    content match {
+      case Success(NonEmptyContent(row, partitionKey)) =>
+        EventTransformer.transform(row) match {
+          case Success(s) => NonEmptyContent(s, partitionKey).success
+          case Failure(f) => new IllegalArgumentException(f.head.toString).failureNel
+        }
+      case Success(FilteredContent) => FilteredContent.success
+      case Failure(f) => throw new IllegalStateException(s"Preceding operation has failed '$content': ${f.head}")
```

Comment:
You're right, the assumption was any failures are from a preceding operator where we return a failureNel with one item.